services/brig/src/Brig/API/Public.hs

```diff
@@ -486,57 +432,56 @@ apiDocs =
-  pure k
 
 -- | Parse a 'PropertyValue' from a bytestring.  This is different from 'FromJSON' in that
 -- checks the byte size of the input, and fails *without consuming all of it* if that size
 -- exceeds the settings.
-safeParsePropertyValue :: IO Lazy.ByteString -> (Handler r) Public.PropertyValue
-safeParsePropertyValue lreqbody = do
+safeParsePropertyValue :: Public.RawPropertyValue -> Handler r Public.PropertyValue
+safeParsePropertyValue raw = do
   maxValueLen <- fromMaybe defMaxValueLen <$> view (settings . propertyMaxValueLen)
-  lbs <- Lazy.take (maxValueLen + 1) <$> liftIO lreqbody
+  let lbs = Lazy.take (maxValueLen + 1) (Public.rawPropertyBytes raw)
   unless (Lazy.length lbs <= maxValueLen) $
     throwStd propertyValueTooLarge
-  hoistEither $ fmapL (StdError . badRequest . pack) (eitherDecode lbs)
+  hoistEither $ first (StdError . badRequest . pack) (propertyValueFromRaw raw)
+
+propertyValueFromRaw :: Public.RawPropertyValue -> Either String Public.PropertyValue
+propertyValueFromRaw raw =
+  Public.PropertyValue raw
+    <$> eitherDecode (Public.rawPropertyBytes raw)
 
-deletePropertyH :: UserId ::: ConnId ::: Public.PropertyKey -> (Handler r) Response
-deletePropertyH (u ::: c ::: k) = lift (API.deleteProperty u c k) >> return empty
+deleteProperty :: UserId -> ConnId -> Public.PropertyKey -> Handler r ()
+deleteProperty u c k = lift (API.deleteProperty u c k)
 
-clearPropertiesH :: UserId ::: ConnId -> (Handler r) Response
-clearPropertiesH (u ::: c) = lift (API.clearProperties u c) >> return empty
+clearProperties :: UserId -> ConnId -> Handler r ()
+clearProperties u c = lift (API.clearProperties u c)
 
-getPropertyH :: UserId ::: Public.PropertyKey ::: JSON -> (Handler r) Response
-getPropertyH (u ::: k ::: _) = do
-  val <- lift . wrapClient $ API.lookupProperty u k
-  return $ case val of
-    Nothing -> setStatus status404 empty
-    Just v -> json (v :: Public.PropertyValue)
+getProperty :: UserId -> Public.PropertyKey -> Handler r (Maybe Public.RawPropertyValue)
+getProperty u k = lift . wrapClient $ API.lookupProperty u k
 
-listPropertyKeysH :: UserId ::: JSON -> (Handler r) Response
-listPropertyKeysH (u ::: _) = do
-  keys <- lift $ wrapClient (API.lookupPropertyKeys u)
-  pure $ json (keys :: [Public.PropertyKey])
+listPropertyKeys :: UserId -> Handler r [Public.PropertyKey]
+listPropertyKeys u = lift $ wrapClient (API.lookupPropertyKeys u)
 
-listPropertyKeysAndValuesH :: UserId ::: JSON -> (Handler r) Response
-listPropertyKeysAndValuesH (u ::: _) = do
-  keysAndVals <- lift $ wrapClient (API.lookupPropertyKeysAndValues u)
-  pure $ json (keysAndVals :: Public.PropertyKeysAndValues)
+listPropertyKeysAndValues :: UserId -> Handler r Public.PropertyKeysAndValues
+listPropertyKeysAndValues u = do
+  keysAndVals <- fmap Map.fromList . lift $ wrapClient (API.lookupPropertyKeysAndValues u)
+  fmap Public.PropertyKeysAndValues . hoistEither $
+    traverse
+      ( first (const (StdError internalServerError))
```

Comment:
Bit ugly to explicitly throw a 500, but I guess this is what was happening earlier when Cql parsing failed? Did you check if the logging is similar?

Reply:
Hey, I'll take a look.

Reply:
> This can only happen if the database is corrupted, so 500 seems appropriate. What's your reservation about it?

5xx errors cause alarms, so I expect us to try to avoid them whenever possible. But in this case I guess it is fair to throw this, which looked like we're throwing 500 on an "expected" error. As I write this I realize my reservation wasn't very well founded, so please ignore me on this.

> Since it feels slightly dangerous to show the details of the parse error to the client in this case, I went with just explicitly logging the error when this happens, and still return the opaque 500 response. Does that sound reasonable? Do you have other suggestions?

Sounds good to me.

> The information is all there, but it seems the error gets logged twice. I suspect this is due to the polysemy refactoring, probably something around the Final IO effect ends up duplicating some logging action, or something along these lines.

Sounds a like a separate issue, I will mark this comment as resolved, please track it separately.

Reply:
> Here is the log output before:
> 
> ```
> [brig] E, request=N/A, cql-io: protocol error: parse error: response body reading: Failed reading: Failed to read property value: Error in $: Failed reading: not a valid json value at 'garbagebaz' Empty call stack
> [brig] D, code=500, label=server-error, request=N/A, "Server Error"
> ```
> 
> and here it is now:
> 
> ```
> [brig] E, request=N/A, Failed to parse a stored property value, raw_value="baz", parse_error=Error in $: Failed reading: not a valid json value at 'garbagebaz'
> [brig] D, code=500, label=internal-server-error, request=N/A, "Internal Server Error"
> [brig] E, request=N/A, Error {code = Status {statusCode = 500, statusMessage = "Internal Server Error"}, label = "internal-server-error", message = "Internal Server Error", errorData = Nothing}
> [brig] D, code=500, label=internal-server-error, request=N/A, "Internal Server Error"
> ```
> 
> The information is all there, but it seems the error gets logged twice. I suspect this is due to the polysemy refactoring, probably something around the `Final IO` effect ends up duplicating some logging action, or something along these lines.
> 
> @mdimjasevic Could you maybe take a look at some point?


What exactly is logged twice? There are four lines in that second snippet and I guess the lines correspond to two pairs of successive errors. The first error line (marked with E) is not the same as the second error line, so it doesn't seem strange to me that each is paired with its own debug line (marked with D). It could be I am missing something so please provide more context.